mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala

```diff
@@ -350,6 +458,145 @@ object GaussianMixture extends DefaultParamsReadable[GaussianMixture] {
 
   @Since("2.0.0")
   override def load(path: String): GaussianMixture = super.load(path)
+
+  /**
+   * Heuristic to distribute the computation of the [[MultivariateGaussian]]s, approximately when
+   * d > 25 except for when k is very small.
+   *
+   * @param k  Number of topics
+   * @param d  Number of features
+   */
+  private[clustering] def shouldDistributeGaussians(k: Int, d: Int): Boolean = {
+    ((k - 1.0) / k) * d > 25
+  }
+
+  /**
+   * Unpack upper triangular part of a symmetric matrix.
+   * @param n The order of the n by n matrix.
+   * @param triangular The upper triangular part of the matrix packed in an array (column major).
+   * @return An array which represents the symmetric matrix in column major.
+   */
+  private[clustering] def unpackUpperTriangularMatrix(
+      n: Int,
+      triangular: Array[Double]): Array[Double] = {
+    val symmetric = Array.fill(n * n)(0.0)
+    var r = 0
+    for (i <- 0 until n) {
+      for (j <- 0 to i) {
+        symmetric(i * n + j) = triangular(r)
+        symmetric(j * n + i) = triangular(r)
+        r += 1
+      }
+    }
+    symmetric
+  }
+
+  private[clustering] def updateWeightsAndGaussians(
+      mean: DenseVector,
+      cov: DenseVector,
+      weight: Double,
+      sumWeights: Double): (Double, (DenseVector, DenseVector)) = {
+    BLAS.scal(1.0 / weight, mean)
+    BLAS.spr(-weight, mean, cov)
+    BLAS.scal(1.0 / weight, cov)
+    val newWeight = weight / sumWeights
+    val newGaussian = (mean, cov)
+    (newWeight, newGaussian)
+  }
+}
+
+/**
+ * ExpectationAggregator computes the partial expectation results.
```

Comment:
We have typically used this documentation as a place to explain the math used to compute updates. It would be nice to have that here as well.